File: day5/group_anagrams.cpp
class Solution {
public:
  vector<vector<string>> groupAnagrams(vector<string> &strs) {
    unordered_map<string, vector<string>> seen;
    int n = strs.size();
    for (int x = 0; x < n; ++x) {
      string copy = strs[x];
      sort(copy.begin(), copy.end());
      seen[copy].push_back(strs[x]);
    }

    vector<vector<string>> ans;

    for (auto i : seen) {
      ans.push_back((i).second);
    }

    return ans;
  }
};

int main(void) {}

File: archives/april152023/group_anagrams.py
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        letter_map = {}
        seen = {}

        for anagram in strs:
            for letter in anagram:
                if letter not in letter_map:
                    letter_map[letter] = [anagram]
                else:
                    letter_map[letter].append(anagram)

        ans = []

        for anagram in strs:
            if len(anagram) > 0:
                starting_set = letter_map[anagram[0]]
                remainder = {associate: True for associate in starting_set}

                for index, letter in enumerate(anagram, 1):
                    associated_anagrams = letter_map[letter]
                    # for each remainder, see if they still exist in the current iteration
                    for key in list(remainder):
                        if key not in associated_anagrams or key in seen:
                            del remainder[key]
            else:
                ans.append([""])

            if len(remainder) > 0:
                ans.append([key for key in remainder])
                seen.update({key: True for key in remainder})
        return ans


anagrams = ["eat", "tea", "tan", "ate", "nat", "bat"]
s = Solution()
print(s.groupAnagrams(anagrams))
